### PR TITLE
Release notes automation

### LINKED ISSUE #4009: Automate consolidated release notes creation
### Is your feature request related to a problem? Please describe

Right now we are using a temporary script https://github.com/opensearch-project/opensearch-build/blob/main/scripts/release-notes/release-notes.sh to consolidate release notes. Its is a very painful process to keep create [consolidated release notes ](https://github.com/opensearch-project/opensearch-build/blob/main/release-notes/opensearch-release-notes-2.9.0.md) for each release and manually keep making change.

### Describe the solution you'd like

We would like to create release notes generation workflow integrated with the python code base that would take care of collecting all plugins' components release notes and creating a consolidate one as per the [standards](https://github.com/opensearch-project/opensearch-plugins/blob/main/RELEASE_NOTES.md).

### Describe alternatives you've considered

Keep doing it manually

### Acceptance Criteria

* The release notes generation should be automated completely along with formatting.
* Only manual thing to do would be adding release highlights, experimental features, etc
* ~Throw a warning if the plugin release notes do not align with the [standards](https://github.com/opensearch-project/opensearch-plugins/blob/main/RELEASE_NOTES.md)~ Assume that all components' release notes are up to the standards.
* Code should also contain tests cases testing various scenarios.

### PR DESCRIPTION
**Pull request by SachinSahu431**
### Description
Python script to automate the process of collecting all plugins' components release notes and creating a consolidated release note.

### Issues Resolved
Closes #4009

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peterzhuamazon** [comment]
Reviewed in officehour, suggest @SachinSahu431 to sort the repos based on repo name before adding entries.
Once tests are added we can start reviewing this before adding github actions.

Thanks.

**Review by peterzhuamazon — COMMENTED** [review]
Thanks @SachinSahu431 for the initial draft, ping @gaiksaya for the co-review of this PR.

Two questions:
1. Could you only update new package in pipfile.lock instead of touching others? I remember able to do that in the past but I dont recall the commands I use. Just to make sure we are not upgrading the existing stable versions of other packages.
2. You have a lot of .md and .txt in the automation dir, are they needed or just for testing purposes?
3. Considering adding some sample commands and descriptions to README file within the new dir you created, see build_workflow for example but this can be done later.
4. Can this be part of the existing `release_notes_workflow` dir?


Thanks.

**Comment by SachinSahu431** [comment]
@peterzhuamazon Thank you for the review!

1. I've undone the modifications in Pipfile.lock and updated it for new package with  `pipenv lock --keep-outdated`.
2. Affirming that the `.md` and `.txt` files are specifically used to validate the results of the ongoing Python script; they will be removed before the merge.
3. The directory has been reorganized, and the `README.md` file has been updated.

Thanks again!

File: src/release_notes_workflow/README.md [hunk 1]
```diff
@@ -1,13 +1,19 @@
-#### Components Release Notes Check 
+- [Consolidated Release Notes Automation](#consolidated-release-notes-automation)
+  - [Components Release Notes Check](#components-release-notes-check)
+  - [Consolidated Release Notes](#consolidated-release-notes)
+
+## Consolidated Release Notes Automation
```
Comment (gaiksaya) [inline src/release_notes_workflow/README.md:5]: nit
```suggestion
## Consolidated Release Notes
```

**Review by gaiksaya — COMMENTED** [review]
Thanks for the contribution @SachinSahu431 . This will definitely save us a lot of manual effort. 
Added some nits and questions. Few suggestions:

1. Please replace all print statements with `logging.info` as best practice. 
2. Remove all temp / test files
3. Fix test cases and add new ones for new methods.

File: src/run_releasenotes_check.py [hunk 2]
```diff
@@ -15,12 +23,181 @@ def main() -> int:
...
+    BASE_FILE_PATH = "release_notes_workflow/results"
+    table_filename = f"{BASE_FILE_PATH}/release_notes_table-{BUILD_VERSION}.md"
+    urls_filename = f"{BASE_FILE_PATH}/release_notes_urls-{BUILD_VERSION}.txt"
+
+    def capitalize_acronyms(formatted_name: str) -> str:
+        acronyms = ["sql", "ml", "knn"]
+        for acronym in acronyms:
+            pattern = re.compile(re.escape(acronym), re.IGNORECASE)
+            formatted_name = re.sub(pattern, acronym.upper(), formatted_name)
+        return formatted_name
```
[the remark this answers is not in the document]
Comment (SachinSahu431) [reply src/run_releasenotes_check.py:38]: I've updated the code that will convert names as follows: 
"knn": "k-NN", 
"k-nn": "k-NN", 
"kNN": "k-NN", 
"ml-commons": "ML Commons"

These names I've taken after taking reference from existing release notes [2.10](https://github.com/opensearch-project/opensearch-build/blob/main/release-notes/opensearch-release-notes-2.10.0.md#opensearch-k-nn), [2.11](https://github.com/opensearch-project/opensearch-build/blob/main/release-notes/opensearch-release-notes-2.11.0.md#opensearch-ml-commons). Please let me know if some changes are required.

File: src/run_releasenotes_check.py [hunk 3]
```diff
@@ -15,12 +23,181 @@ def main() -> int:
...
-    release_notes = ReleaseNotes(manifest_file, args.date)
-    if args.action == "check":
-        table_output = release_notes.table()
-        table_output.write_table()
+    BUILD_VERSION = manifest_file.build.version
+
+    # storing temporary release notes for testing purposes
+    BASE_FILE_PATH = "release_notes_workflow/results"
+    table_filename = f"{BASE_FILE_PATH}/release_notes_table-{BUILD_VERSION}.md"
+    urls_filename = f"{BASE_FILE_PATH}/release_notes_urls-{BUILD_VERSION}.txt"
```
[the remark this answers is not in the document]
Comment (SachinSahu431) [reply src/run_releasenotes_check.py:31]: I'm adding them under `.gitignore`

**Comment by gaiksaya** [comment]
Hi @SachinSahu431 

Please feel free to convert this PR ready for review when you are ready. Looks like tests needs to be fixed and also some comments need to be addressed.
Thanks!

**Comment by SachinSahu431** [comment]
Hi @gaiksaya!

I have added tests for the new `compile` option. Explicit tests for the `check` option are not required as it is a subset of the compile option. Also, since the consolidation script ignores `OpenSearch` component, I've created a new manifest file with component named `Opensearch-test` in `manifests/templates/opensearch/test` for testing purposes.

File: manifests/templates/opensearch/test/manifest.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+---
```
Comment (gaiksaya) [inline manifests/templates/opensearch/test/manifest.yml:1]: Any test data that you need to add can be added under https://github.com/opensearch-project/opensearch-build/tree/main/tests/data

Comment (SachinSahu431) [reply manifests/templates/opensearch/test/manifest.yml]: I have moved the test data, thanks!

**Comment by bbarani** [comment]
@gaiksaya @peterzhuamazon Can we move ahead with this PR?

**Review by gaiksaya — COMMENTED** [review]
Looks like code coverage as commented on the lines missing the test cases. Is it possible for you to add those?

If not, you can take that up as a follow up PR.

File: .gitignore [hunk 5]
```diff
@@ -37,3 +37,4 @@ out.txt
 vars/
 
 test-report.yml
+src/release_notes_workflow/results/*
```
[the remark this answers is not in the document]
Comment (SachinSahu431) [reply .gitignore:40]: @gaiksaya Yes, this is to ignore the folder containing temporary release notes created by script.

[the remark this answers is not in the document]
Comment (SachinSahu431) [reply .gitignore:40]: We can work on missing test cases in a new PR.

**Review by gaiksaya — APPROVED** [review]

**Pull request merged**